nautobot/utilities/filters.py

```diff
@@ -286,6 +287,105 @@ def filter(self, qs, value):
         return qs
 
 
+class MappedPredicatesFilterMixin:
+    """
+    A filter mixin to provide the ability to specify fields and lookup
+    expressions to use for filtering.
+
+    A mapping of filter predicates (field_name: lookup_expr) must be provided to
+    the filter when declared on a filterset. This mapping is used to construct a
+    `Q` query to filter based on the provided predicates.
+    """
+
+    label = None
+
+    def __init__(self, filter_predicates=None, distinct=True, label=None, *args, **kwargs):
+        if not isinstance(filter_predicates, dict):
+            raise TypeError("filter_predicates must be a dict")
+
+        # Format: field_name: lookup_expr
+        self.filter_predicates = filter_predicates
+
+        # Try to use the label from the class if it is defined.
+        if label is None:
+            kwargs.setdefault("label", self.label)
+
+        kwargs.setdefault("distinct", distinct)
+        super().__init__(*args, **kwargs)
+
+    def filter(self, qs, value):
+        if value in EMPTY_VALUES:
+            return qs
+
+        query = models.Q()
+        value = value.strip()
+        for field_name, lookup_expr in self.filter_predicates.items():
+            # Try to capture the optional type (otherwise we assume str)
+            if ":" in field_name:
+                field_name, field_type = field_name.split(":")
+            else:
+                field_type = None
+
+            # Try to cast value to the specified type, or skip creating a
+            # predicate for it.
+            if field_type is not None:
+                try:
+                    type_constructor = getattr(builtins, field_type)
+                except AttributeError as err:
+                    raise TypeError(f"Invalid type {field_type} for {field_name}") from err
+
+                try:
+                    value = type_constructor(value)
+                except (TypeError, ValueError):
+                    continue
+
+            predicate = {f"{field_name}__{lookup_expr}": value}
+            query |= models.Q(**predicate)
+
+        if self.distinct:
+            qs = qs.distinct()
```

Comment:
Can you think of a case where we *shouldn't* use `distinct` on a filter? I wonder if we should remove the option here and just make it always-on.

Reply:
Since we don't do any "number of times matched" calculation, I would agree.

Reply:
That was my thought as well.